### PR TITLE
New Azure Batch DataPlane SDK

### PR DESCRIPTION
**Pull request by bgklein**

### CODE REVIEW

File: sdk/batch/Microsoft.Azure.Batch/Tools/ObjectModelCodeGenerator/Spec/CloudJob.json [hunk 1]
```diff
@@ -112,7 +112,7 @@
...
+            "SummaryComment": "The Job Release Task runs when the Job ends, because of one of the following: The user calls the Terminate Job API, or the Delete Job API while the Job is still active, the Job's maximum wall clock time constraint is reached, and the Job is still active, or the Job's Job Manager Task completed, and the Job is configured to terminate when the Job Manager completes. The Job Release Task runs on each Node where Tasks of the Job have run and the Job Preparation Task ran and completed. If you reimage a Node after it has run the Job Preparation Task, and the Job ends without any further Tasks of the Job running on that Node (and hence the Job Preparation Task does not re-run), then the Job Release Task does not run on that Compute Node. If a Node reboots while the Job Release Task is still running, the Job Release Task runs again when the Compute Node starts up. The Job is not marked as complete until all Job Release Tasks have completed. The Job Release Task runs in the background. It does not occupy a scheduling slot; that is, it does not count towards the taskSlotsPerNode limit specified on the Pool.",
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/Tools/ObjectModelCodeGenerator/Spec/CloudJob.json:115]: This is difficult to follow. Should we maybe just link to some documentation?

File: sdk/batch/Microsoft.Azure.Batch/Tools/ObjectModelCodeGenerator/Spec/CloudTask.json [hunk 2]
```diff
@@ -336,6 +336,17 @@
         "UnboundAccess": "read,write"
       },
       "Value": null
+    },
+    {
+      "Key": {
+        "Type": "int?",
+        "Name": "RequiredSlots",
+        "SummaryComment": "The number of scheduling slots that the Task required to run.",
+        "RemarksComment": "The default is 1. A Task can only be scheduled to run on a compute node if the node has enough free scheduling slots available. For multi-instance Tasks, this must be 1.",
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/Tools/ObjectModelCodeGenerator/Spec/CloudTask.json:345]: What if it's not 1? Does it default to 1? Will it fail?

Comment (bgklein) [reply sdk/batch/Microsoft.Azure.Batch/Tools/ObjectModelCodeGenerator/Spec/CloudTask.json:345]: The task add will fail once it hits the FE

File: sdk/batch/Microsoft.Azure.Batch/Tools/ObjectModelCodeGenerator/Spec/CloudJob.json (hunk 1 above)
Comment (bgklein) [reply sdk/batch/Microsoft.Azure.Batch/Tools/ObjectModelCodeGenerator/Spec/CloudJob.json:115]: This is from the swagger definition. There are some problems with hardlinking external documentation, so generally this is meant as a summary and not full clarity.

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/DataDisk.cs [hunk 3]
```diff
@@ -57,7 +57,8 @@ public DataDisk(int lun, int diskSizeGB, CachingType? caching = default(CachingT
         /// </summary>
         /// <remarks>
         /// The lun is used to uniquely identify each data disk. If attaching
-        /// multiple disks, each should have a distinct lun.
+        /// multiple disks, each should have a distinct lun. The value must be
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/DataDisk.cs:60]: Can we specify what "lun" stands for or means?

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/JobManagerTask.cs [hunk 4]
```diff
@@ -89,7 +91,7 @@ public JobManagerTask()
...
+        public JobManagerTask(string id, string commandLine, string displayName = default(string), TaskContainerSettings containerSettings = default(TaskContainerSettings), IList<ResourceFile> resourceFiles = default(IList<ResourceFile>), IList<OutputFile> outputFiles = default(IList<OutputFile>), IList<EnvironmentSetting> environmentSettings = default(IList<EnvironmentSetting>), TaskConstraints constraints = default(TaskConstraints), int? requiredSlots = default(int?), bool? killJobOnCompletion = default(bool?), UserIdentity userIdentity = default(UserIdentity), bool? runExclusive = default(bool?), IList<ApplicationPackageReference> applicationPackageReferences = default(IList<ApplicationPackageReference>), AuthenticationTokenSettings authenticationTokenSettings = default(AuthenticationTokenSettings), bool? allowLowPriorityNode = default(bool?))
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/JobManagerTask.cs:94]: We should create a class/struct (e.g. JobManagerTaskOptions) to hold the majority of these parameters, and the user just passes that in (along with maybe the id and commandLine). Can be pushed to a separate PR, to not hold this one up.

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolEndpointConfiguration.cs [hunk 5]
```diff
@@ -52,7 +52,8 @@ public PoolEndpointConfiguration(IList<InboundNATPool> inboundNATPools)
         /// <remarks>
         /// The maximum number of inbound NAT Pools per Batch Pool is 5. If the
         /// maximum number of inbound NAT Pools is exceeded the request fails
-        /// with HTTP status code 400.
+        /// with HTTP status code 400. This cannot be specified if the
+        /// IPAddressProvisioningType is NoPublicIPAddresses.
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolEndpointConfiguration.cs:56]: Is IPAddressProvisioningType an enum we can link to?

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolSpecification.cs [hunk 6]
```diff
@@ -75,13 +76,13 @@ public PoolSpecification()
...
...iguration = default(CloudServiceConfiguration), VirtualMachineConfiguration virtualMachineConfiguration = default(VirtualMachineConfiguration), int? taskSlotsPerNode = default(int?), TaskSchedulingPolicy taskSchedulingPolicy = default(TaskSchedulingPolicy), System.TimeSpan? resizeTimeout = default(System.TimeSpan?), int? targetDedicatedNodes = default(int?), int? targetLowPriorityNodes = default(int?), bool? enableAutoScale = default(bool?), string autoScaleFormula = default(string), System.TimeSpan? autoScaleEvaluationInterval = default(System.TimeSpan?), bool? enableInterNodeCommunication = default(bool?), NetworkConfiguration networkConfiguration = default(NetworkConfiguration), StartTask startTask = default(StartTask), IList<CertificateReference> certificateReferences = default(IList<CertificateReference>), IList<ApplicationPackageReference> applicationPackageReferences = default(IList<ApplicationPackageReference>), IList<string> applicationLicenses = default(IList<string>), IList<UserAccount> userAccounts = default(IList<UserAccount>), IList<MetadataItem> metadata = default(IList<MetadataItem>), IList<MountConfiguration> mountConfiguration = default(IList<MountConfiguration>))
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolSpecification.cs:79]: Another candidates for an options argument. I won't comment on any others to avoid spamming.

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolSpecification.cs [hunk 7]
```diff
@@ -159,15 +160,15 @@ public PoolSpecification(string vmSize, string displayName = default(string), Cl
...
         /// <summary>
-        /// Gets or sets the maximum number of Tasks that can run concurrently
-        /// on a single Compute Node in the Pool.
+        /// Gets or sets the number of task slots that can be used to run
+        /// concurrent tasks on a single compute node in the pool.
         /// </summary>
         /// <remarks>
         /// The default value is 1. The maximum value is the smaller of 4 times
-        /// the number of cores of the vmSize of the Pool or 256.
+        /// the number of cores of the vmSize of the pool or 256.
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolSpecification.cs:168]: We can govern this in the setter.

File: sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs [hunk 8]
```diff
@@ -939,7 +939,60 @@ public async Task Job_GetTaskCounts_ReturnsCorrectCount()
 
                         var counts = await unboundJob.GetTaskCountsAsync().ConfigureAwait(false);
 
-                        Assert.Equal(2, counts.Active);
+                        Assert.Equal(2, counts.TaskCounts.Active);
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs:942]: Are we testing two different things here? Not sure what the semantic difference is, but if it is two different things, we should probably break it into two different tests.

File: sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs [hunk 9]
```diff
@@ -939,7 +939,60 @@ public async Task Job_GetTaskCounts_ReturnsCorrectCount()
...
+        [Fact]
+        [LiveTest]
+        [Trait(TestTraits.Duration.TraitName, TestTraits.Duration.Values.ShortDuration)]
+        public async Task Job_GetTaskCounts_ReturnsCorrectCountNonZeroTaskSlots()
+        {
+            Func<Task> test = async () =>
+            {
+                using (BatchClient batchCli = await TestUtilities.OpenBatchClientFromEnvironmentAsync().ConfigureAwait(false))
+                {
+                    string jobId = "NonZeroTaskSlots-" + TestUtilities.GetMyName();
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs:965]: nit: consider string interpolation over concatenation.

File: sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs [hunk 10]
```diff
@@ -939,7 +939,60 @@ public async Task Job_GetTaskCounts_ReturnsCorrectCount()
...
+                        await unboundJob.RefreshAsync().ConfigureAwait(false);
+
+                        CloudTask t1 = new CloudTask("t1", "cmd /c dir");
+                        t1.RequiredSlots = 2;
+                        CloudTask t2 = new CloudTask("t2", "cmd /c ping 127.0.0.1 -n 4");
+                        t2.RequiredSlots = 3;
+
+                        await unboundJob.AddTaskAsync(new[] { t1, t2 }).ConfigureAwait(false);
+
+                        await Task.Delay(TimeSpan.FromSeconds(5)).ConfigureAwait(false); // Give the service some time to get the counts
```
Comment (paterasMSFT) [inline sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs:984]: Relying on time might lead to a flaky test. Is there something we can poll or wait on, instead?

**Review by paterasMSFT — CHANGES_REQUESTED** [review]
Looks good! I left a few comments, but they're mostly minor. The only one of potential concern is this one: https://github.com/Azure/azure-sdk-for-net/pull/15061#pullrequestreview-494917896

**Comment by paterasMSFT** [comment]
BTW, I love how you not only keep the documentation current, but also expand on it. That's really above and beyond, and provides a lot of value!

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/DataDisk.cs (hunk 3 above)
Comment (bgklein) [reply sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/DataDisk.cs:60]: This model is purely autogenerated from swagger. Generally speaking it is best to update the actual Swagger defined RestAPI documentation, because any change here will just get replaced each generation.

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolEndpointConfiguration.cs (hunk 5 above)
Comment (bgklein) [reply sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolEndpointConfiguration.cs:56]: This model is purely autogenerated from swagger. Generally speaking it is best to update the actual Swagger defined RestAPI documentation, because any change here will just get replaced each generation.

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolSpecification.cs (hunk 7 above)
Comment (bgklein) [reply sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/PoolSpecification.cs:168]: This model is purely autogenerated from swagger. Generally speaking it is best to update the actual Swagger defined RestAPI documentation, because any change here will just get replaced each generation.

File: sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/DataDisk.cs (hunk 3 above)
Comment (paterasMSFT) [reply sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/DataDisk.cs:60]: Right. Can we not do it there? Or did someone else update the docs and this PR is just picking up on it?

Comment (paterasMSFT) [reply sdk/batch/Microsoft.Azure.Batch/src/GeneratedProtocol/Models/DataDisk.cs:60]: I guess it's the latter, since those files aren't in the change long here. Fair enough.

File: sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs (hunk 10 above)
Comment (bgklein) [reply sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs:984]: Unfortunately not. This part of the service is actually itself flaky so our best time is time wait

File: sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs (hunk 9 above)
Comment (paterasMSFT) [reply sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs:965]: Assuming this was resolved as a will not fix?

File: sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs (hunk 10 above)
Comment (paterasMSFT) [reply sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/CloudJobIntegrationTests.cs:984]: Let's discuss this on a call.

**Review by paterasMSFT — APPROVED** [review]
